accessibility/src/observer.rs

```diff
@@ -0,0 +1,194 @@
+use std::{ffi::c_void, str::FromStr};
+
+use accessibility_sys::{
+    pid_t, AXObserverAddNotification, AXObserverCallback, AXObserverCallbackWithInfo,
+    AXObserverCreate, AXObserverCreateWithInfoCallback, AXObserverGetRunLoopSource,
+    AXObserverGetTypeID, AXObserverRef, AXObserverRemoveNotification, AXUIElementRef,
+};
+
+use crate::{
+    util::{ax_call, ax_call_void},
+    Error,
+};
+
+use core_foundation::{
+    base::TCFType,
+    declare_TCFType, impl_CFTypeDescription, impl_TCFType,
+    runloop::{kCFRunLoopDefaultMode, CFRunLoop},
+    string::CFString,
+};
+
+use super::AXUIElement;
+
+declare_TCFType!(AXObserver, AXObserverRef);
+impl_TCFType!(AXObserver, AXObserverRef, AXObserverGetTypeID);
+impl_CFTypeDescription!(AXObserver);
+
+unsafe impl Send for AXObserver {}
+
+impl AXObserver {
+    pub fn new(pid: pid_t, callback: AXObserverCallback) -> Result<Self, Error> {
+        unsafe {
+            Ok(TCFType::wrap_under_create_rule(
+                ax_call(|x| AXObserverCreate(pid, callback, x)).map_err(Error::Ax)?,
+            ))
+        }
+    }
+
+    pub fn new_with_info(pid: pid_t, callback: AXObserverCallbackWithInfo) -> Result<Self, Error> {
+        unsafe {
+            Ok(TCFType::wrap_under_create_rule(
+                ax_call(|x| AXObserverCreateWithInfoCallback(pid, callback, x))
+                    .map_err(Error::Ax)?,
+            ))
+        }
+    }
+
+    pub fn new_from_bundle(bundle_id: &str, callback: AXObserverCallback) -> Result<Self, Error> {
+        let bundle_ui_element = AXUIElement::application_with_bundle(bundle_id)?;
+        let bundle_pid = bundle_ui_element.pid()?;
+        unsafe {
+            Ok(TCFType::wrap_under_create_rule(
+                ax_call(|x| AXObserverCreate(bundle_pid, callback, x)).map_err(Error::Ax)?,
+            ))
+        }
+    }
+
+    pub fn new_from_bundle_with_info(
+        bundle_id: &str,
+        callback: AXObserverCallbackWithInfo,
+    ) -> Result<Self, Error> {
+        let bundle_ui_element = AXUIElement::application_with_bundle(bundle_id)?;
+        let bundle_pid = bundle_ui_element.pid()?;
+        unsafe {
+            Ok(TCFType::wrap_under_create_rule(
+                ax_call(|x| AXObserverCreateWithInfoCallback(bundle_pid, callback, x))
+                    .map_err(Error::Ax)?,
+            ))
+        }
+    }
+
+    pub fn add_notification<T>(
+        &mut self,
+        notification: &str,
+        ui_element: &AXUIElement,
+        mut ctx: T,
+    ) -> Result<(), Error> {
+        unsafe {
+            // Create CFStringRef from notification string
+            let notification_cfstr = CFString::from_str(notification).unwrap();
+
+            Ok(ax_call_void(|| {
+                AXObserverAddNotification(
+                    self.0,
+                    ui_element.as_CFTypeRef() as AXUIElementRef,
+                    notification_cfstr.as_concrete_TypeRef(),
+                    &mut ctx as *mut _ as *mut c_void,
```

Comment:
This doesn't look safe to me. It's creating a raw pointer to a T that immediately goes out of scope, no?